### DISCUSSION TITLE
Moving disks with chunks between chunkservers

### DISCUSSION THREAD

**Discussion by Blackpaw** [moosefs/moosefs#425] [Q&A]
Can a disk and its chunks just be moved to another chunkserver? i.e.

**Chunkservers CS1 & CS2**
- Remove Disk from CS1 mfshdd.cfg
- Reload CS2 chunkserver service
- Shut down CS1
- physically swap disk to CS2
- Add disk to CS2 mfshdd.cfg
- Reload CS2 chunkserver service

CS2 Should now be serving up the chunks?

**Comment by jkiebzak** [comment]
Does CS2 have any existing chunks before this disk move? If so, any
duplicate chunks would be detected and removed when you reload
mfschunkserver on cs2. This would then lead to data be undergoal, and then
replication would then kick in.

**Reply by Blackpaw** [reply]
CS2 in a new empty server, CS1 is being retired.

**Reply by oszafraniec** [reply]
@jkiebzak is 100% correct. CS2 will see overgoal chunks and will delete one of them during drive scan. There can be only 1 chunk copy per chunkserver.
If CS1 is being retired and you need only to swap drives to new CS2 chassis, just do it. MFS on CS2 will always scan drives after start, will find chunks on them, report them to master and you're good to go. Remember to have same user/uid/permissions/etc on all CS to make such swaps.

**Reply by Blackpaw** [reply]
Yup, moved all three drives last night, worked a charm.

Thanks!

**Reply by coderfi** [reply]
On this idea, for 'fun', could we just administratively copy (i.e. with rsync or scp) a file such as `chunk_0000000001C37208_00000001.mfs` to every other chunkserver volume? These would effectively become duplicated and removed, right?
Does it matter which sub-folder they reside (e.g. 00/, 01/, 02/, ... FF/)?

**Reply by oszafraniec** [reply]
Yes, chsrv will rescan drives after restart, will see moved/copied chunks, report to master and master will see overgoals and will decide to delete them over time. Just do it when a chunkserver is stoped as a rule of thumb.

**Reply by chogata** [reply]
If you do that while the chunkserver is working, you are asking for trouble. If you do that when the chunkserver is not working and you delete the `.chunkdb` file from each volume, nothing bad will happen - the chunkserver will notice duplicates and delete them after some time (if you copy to another drive on the same chunkserver) or report them to master (if you copy to a drive on another chunkserver) and master will issue a delete order for unnecessary copies. If you don't delete `.chunkdb`, the chunkserver should still detect a change based on atimes/mtimes, but then a lot depends on the filesystem you use on the local drives and how much it "optimises" the use of atime/mtime. Some filesystems will behave in a way that will make it impossible for the chunkserver to notice that something was altered after the `.chunkdb` was created. So it's best to remember to delete it if you try stuff like that.
Better yet - don't try it ;)

**Comment by eleaner** [comment]
yup
I believe it should work
I think I even did it in the past

**Reply by Blackpaw** [reply]
Excellent, thanks!

**Comment by jkiebzak** [comment]
If you stop the chunkserver, and delete .chunkdb, then start the chunkserver, it will crawl the disk and eventually come across those duplicate chunks, and mark them as duplicate.

No, it doesn’t matter which folder you put the chunk in

**Comment by coderfi** [comment]
Ah, thanks for the behavioral description of the mysterious chunkdb file!
Yeah I wouldn’t want to mess around with those files unnecessarily... but
stuff happens sometimes, so good to know these things. :)


On Tue, Apr 27, 2021 at 7:38 AM Agata Kruszona-Zawadzka <
***@***.***> wrote:

> If you do that while the chunkserver is working, you are asking for
> trouble. If you do that when the chunkserver is not working and you delete
> the .chunkdb file from each volume, nothing bad will happen - the
> chunkserver will notice duplicates and delete them after some time (if you
> copy to another drive on the same chunkserver) or report them to master (if
> you copy to a drive on another chunkserver) and master will issue a delete
> order for unnecessary copies. If you don't delete .chunkdb, the
> chunkserver should still detect a change based on atimes/mtimes, but then a
> lot depends on the filesystem you use on the local drives and how much it
> "optimises" the use of atime/mtime. Some filesystems will behave in a way
> that will make it impossible for the chunkserver to notice that something
> was altered after the .chunkdb was created. So it's best to remember to
> delete it if you try stuff like that.
> Better yet - don't try it ;)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/moosefs/moosefs/discussions/425#discussioncomment-664689>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDPY7Z3PJEBWOPBHRXMAXLTK3D7RANCNFSM4ZXNDWQA>
> .
>
-- 
Fairiz "Fi" Azizi
